src/main/java/picard/illumina/BarcodeExtractor.java

```diff
@@ -0,0 +1,248 @@
+                            final int maxNoCalls,
+                            final int maxMismatches,
+                            final int minMismatchDelta,
+                            final int minimumBaseQuality,
+                            final DistanceMetric distanceMode) {
+        this.metrics = barcodeToMetrics;
+        this.barcodeLookupMap = barcodeLookupMap;
+        this.barcodesBytes = barcodeByteStrings;
+        this.noMatch = noMatchMetric;
+        this.maxNoCalls = maxNoCalls;
+        this.maxMismatches = maxMismatches;
+        this.minMismatchDelta = minMismatchDelta;
+        this.minimumBaseQuality = minimumBaseQuality;
+        this.distanceMode = distanceMode;
+    }
+    /**
+     * Find the best barcode match for the given read sequence, and accumulate metrics
+     *
+     * NOTE: the returned BarcodeMatch object will contain mismatches mismatchesToSecondBest values that may be
+     * inaccurate as long as the conclusion match/no-match isn't affected. for example, mismatches and mismatchesToSecondBest
+     * may be smaller than their true value if mismatches is truly larger than maxMismatches.
+     * Also, mismatchesToSecondBest might be smaller than its true value if its true value is greater than
+     * mismatches + minMismatchDelta. This is due to an optimization which allows the distance calculation to stop once
+     * the conclusion (Match or no-Match) can be reached.
+     *
+     * @param readSubsequences portion of read containing barcode
+     * @return perfect barcode string, if there was a match within tolerance, or null if not.
+     */
+    BarcodeMatch findBestBarcode(final byte[][] readSubsequences,
+                                 final byte[][] qualityScores) {
+        final boolean canUseLookupTable = areAllQualitiesAboveMinimum(qualityScores, minimumBaseQuality);
+        final ByteString barcodesAsString = new ByteString(readSubsequences);
+        BarcodeMatch match = canUseLookupTable ? barcodeLookupMap.get(barcodesAsString) : null;
+        if (match == null) {
+            match = calculateBarcodeMatch(readSubsequences, qualityScores);
+
+            if (canUseLookupTable && match.isMatched()) {
+                barcodeLookupMap.put(barcodesAsString, match);
+            }
+        }
+
+        return match;
+    }
+
+    BarcodeMatch calculateBarcodeMatch(final byte[][] readSubsequences,
+                                       final byte[][] qualityScores) {
+        final BarcodeMatch match;
+        String bestBarcode = null;
+        match = new BarcodeMatch();
+
+        int totalBarcodeReadBases = 0;
+        int numNoCalls = 0; // NoCalls are calculated for all the barcodes combined
+
+        for (final byte[] bc : readSubsequences) {
+            totalBarcodeReadBases += bc.length;
+            for (final byte b : bc) {
+                if (SequenceUtil.isNoCall(b)) {
+                    ++numNoCalls;
+                }
+            }
+        }
+
```

Comment:
> I guess it would be a functional change, but I think it would be nice to short-circuit here when `numNoCalls > maxNoCalls` instead of wasting cycles doing all the matching. The only difference would be in the `s_x_barcode.txt` files that are output, not in the BAMs/fastqs.

We could do this for inline matching only and not affect the barcode.txt outputs.